File: 0649-dota2-senate/0649-dota2-senate.java
class Solution {
    public String predictPartyVictory(String senate) {
        Queue<Integer> queueR = new LinkedList();
        Queue<Integer> queueD = new LinkedList();
        for (int i=0; i<senate.length(); i++){
            if (senate.charAt(i) == 'R') {
                queueR.add(i);
            }else{
                queueD.add(i);
            }
        }

        int l = senate.length();

        while(!queueR.isEmpty() && !queueD.isEmpty()){
            int r = queueR.poll();
            int d = queueD.poll();
            
            if (r < d) {
                queueR.add(l + r);
            }else{
                queueD.add(l + d);
            }

        }

        if (!queueR.isEmpty()){
            return "Radiant";
        }else{
            return "Dire";
        }
    }
}

File: 0649-dota2-senate/0649-dota2-senate.go
type Senator struct {
    char rune
    index int
}

func predictPartyVictory(senate string) string {
    rQueue := []Senator{}
    dQueue := []Senator{}

    for i, s := range senate{
        if s == 'R' {
            rQueue = append(rQueue, Senator{index:i, char:s})
        }else{
            dQueue = append(dQueue, Senator{index:i, char:s})
        }
    }

    for len(rQueue) > 0 && len(dQueue) > 0 {
        if rQueue[0].index < dQueue[0].index{
            winner := rQueue[0]
            rQueue = rQueue[1:]
            dQueue = dQueue[1:]
            winner.index += len(senate)
            rQueue = append(rQueue, winner)
        }else{
            winner := dQueue[0]
            rQueue = rQueue[1:]
            dQueue = dQueue[1:]
            winner.index += len(senate)
            dQueue = append(dQueue, winner)
        }
    } 

    if len(rQueue) == 0 {
        return "Dire"
    }else{
        return "Radiant"
    }
}